### PR TITLE
Fix regression with old marshaled specs having null `required_rubygems_version`

### LINKED ISSUE #5290: bundler-2.3.5: undefined method `satisfied_by?' for nil:NilClass
### Describe the problem as clearly as you can

On bundler 2.3.5, in my linux docker container, I see the following error report.  The Gemfile is below and as you can see it uses a single gem, from a third-party gem server.  I've been trying to dig in, and one of the dependencies (pdf-writer) seems to create this nil reference error.  I can't make this happen in my normal Mac environment though, so I don't really understand it beyond this.  There's a Dockerfile down beyond to make it simpler to recreate. If you need to me to try anything else, feel free to suggest things.

Might be related to #4877 

```
NoMethodError: undefined method `satisfied_by?' for nil:NilClass
  /usr/share/gems/gems/bundler-2.3.5/lib/bundler/lazy_specification.rb:95:in `block in __materialize__'
  /usr/share/gems/gems/bundler-2.3.5/lib/bundler/lazy_specification.rb:92:in `select'
  /usr/share/gems/gems/bundler-2.3.5/lib/bundler/lazy_specification.rb:92:in `__materialize__'
  /usr/share/gems/gems/bundler-2.3.5/lib/bundler/spec_set.rb:75:in `block in materialize'
  /usr/share/gems/gems/bundler-2.3.5/lib/bundler/spec_set.rb:72:in `map!'
  /usr/share/gems/gems/bundler-2.3.5/lib/bundler/spec_set.rb:72:in `materialize'
  /usr/share/gems/gems/bundler-2.3.5/lib/bundler/definition.rb:468:in `materialize'
  /usr/share/gems/gems/bundler-2.3.5/lib/bundler/definition.rb:190:in `specs'
  /usr/share/gems/gems/bundler-2.3.5/lib/bundler/installer.rb:243:in `ensure_specs_are_compatible!'
  /usr/share/gems/gems/bundler-2.3.5/lib/bundler/installer.rb:83:in `block in run'
  /usr/share/gems/gems/bundler-2.3.5/lib/bundler/process_lock.rb:12:in `block in lock'
  /usr/share/gems/gems/bundler-2.3.5/lib/bundler/process_lock.rb:9:in `open'
  /usr/share/gems/gems/bundler-2.3.5/lib/bundler/process_lock.rb:9:in `lock'
  /usr/share/gems/gems/bundler-2.3.5/lib/bundler/installer.rb:71:in `run'
  /usr/share/gems/gems/bundler-2.3.5/lib/bundler/installer.rb:23:in `install'
  /usr/share/gems/gems/bundler-2.3.5/lib/bundler/cli/install.rb:62:in `run'
  /usr/share/gems/gems/bundler-2.3.5/lib/bundler/cli.rb:255:in `block in install'
  /usr/share/gems/gems/bundler-2.3.5/lib/bundler/settings.rb:131:in `temporary'
  /usr/share/gems/gems/bundler-2.3.5/lib/bundler/cli.rb:254:in `install'
  /usr/share/gems/gems/bundler-2.3.5/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
  /usr/share/gems/gems/bundler-2.3.5/lib/bundler/vendor/thor/lib/thor/invocation.rb:127:in `invoke_command'
  /usr/share/gems/gems/bundler-2.3.5/lib/bundler/vendor/thor/lib/thor.rb:392:in `dispatch'
  /usr/share/gems/gems/bundler-2.3.5/lib/bundler/cli.rb:31:in `dispatch'
  /usr/share/gems/gems/bundler-2.3.5/lib/bundler/vendor/thor/lib/thor/base.rb:485:in `start'
  /usr/share/gems/gems/bundler-2.3.5/lib/bundler/cli.rb:25:in `start'
  /usr/share/gems/gems/bundler-2.3.5/exe/bundle:48:in `block in <top (required)>'
  /usr/share/gems/gems/bundler-2.3.5/lib/bundler/friendly_errors.rb:103:in `with_friendly_errors'
  /usr/share/gems/gems/bundler-2.3.5/exe/bundle:36:in `<top (required)>'
  /usr/bin/bundle:23:in `load'
  /usr/bin/bundle:23:in `<main>'
```

## Environment

```
Bundler       2.3.5
  Platforms   ruby, x86_64-linux
Ruby          2.7.4p191 (2021-07-07 revision a21a3b7d23704a01d34bd79d09dc37897e00922a) [x86_64-linux]
  Full Path   /usr/bin/ruby
  Config Dir  /etc
RubyGems      3.1.6
  Gem Home    /usr/share/gems
  Gem Path    /root/.gem/ruby:/usr/share/gems:/usr/local/share/gems
  User Home   /root
  User Path   /root/.gem/ruby
  Bin Dir     /usr/bin
OpenSSL
  Compiled    OpenSSL 1.1.1g FIPS  21 Apr 2020
  Loaded      OpenSSL 1.1.1k  FIPS 25 Mar 2021
  Cert File   /etc/pki/tls/cert.pem
  Cert Dir    /etc/pki/tls/certs
Tools
  Git         2.31.1
  RVM         not installed
  rbenv       not installed
  chruby      not installed
```

## Bundler Build Metadata

```
Built At          2022-01-12
Git SHA           a13d015fcb
Released Version  true
```

### Gemfile

```ruby
source "https://rubygems.org"

gem "ruport", "=1.7.0.3", :source => "https://rubygems.manageiq.org"
```

### Gemfile.lock

```
<No /test/Gemfile.lock found>
```

### Did you try upgrading rubygems & bundler?

This uses the latest version of bundler 2.3.5, and rubygems 3.1.6.  I can't seem to update rubygems because `gem update --system` seems to just hang forever.

Note that bundler 2.3.4 works fine.

### Post steps to reproduce the problem

This Dockerfile is the simplest reproducer I could get.  It will fail on the `bundle install` line when you try to `docker build` it.  If you comment out that line, it will drop you into a bash prompt where you can type `bundle install` manually.  You can also change the version of bundler to 2.3.4 to see that it works correctly.

```dockerfile
FROM registry.access.redhat.com/ubi8/ubi:8.4

RUN dnf -y module enable ruby:2.7 && dnf -y install ruby-devel
RUN gem install bundler:2.3.5

WORKDIR "test"

RUN echo -e 'source "https://rubygems.org"\n\n\
gem "ruport", "=1.7.0.3", :source => "https://rubygems.manageiq.org"'\
> Gemfile

RUN bundle install

ENTRYPOINT ["/bin/bash"]
```

### Which command did you run?

`bundle install`

### What were you expecting to happen?

Not to blow up

### What actually happened?

It blew up

### PR DESCRIPTION
**Pull request by deivid-rodriguez**
## What was the end-user or developer problem that led to this PR?

I made [a commit](https://github.com/rubygems/rubygems/commit/ab0a1c16c9e92e175e50e2e386105725a5c7cc87) that failed to consider that some old marshaled specs may have `required_rubygems_version` set to `nil`.

## What is your fix for the problem, implemented in this PR?

Add a fallback for this edge case.

Fixes #5290.

## Make sure the following tasks are checked

- [x] Describe the problem / feature
- [x] Write [tests](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#tests) for features and bug fixes
- [x] Write code to solve the problem
- [x] Make sure you follow the [current code style](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#code-formatting) and [write meaningful commit messages without tags](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#commit-messages)

### CODE REVIEW

File: bundler/spec/realworld/double_check_spec.rb [hunk 1]
```diff
@@ -37,4 +37,12 @@
     ruby cmd
     ruby cmd
   end
+
+  it "deals with old marshaled specs with nil required_rubygems_version" do
+    install_gemfile <<~G
+      source "https://rubygems.org"
+
+      gem "ruport", "=1.7.0.3", :source => "https://rubygems.manageiq.org"
```
Comment (Fryguy) [inline bundler/spec/realworld/double_check_spec.rb:45]: I'm not sure this is a good idea in specs, only because there's a potential it could go away in the future (unless this doesn't actually hit the live system? I see some recorded http things below).  For right now though, it's probably fine.

For testing purposes does rubygems have a third party package registry. If not, maybe Github package registry could be used for that purpose and dedicated to testing rubygems/bundler?

Comment (deivid-rodriguez) [reply bundler/spec/realworld/double_check_spec.rb:45]: @Fryguy Correct, this doesn't hit the live system, it uses the recorded responses. Normally I tried to avoid realworld specs like this one, I only use them when I cannot find another way to test what I want to test. In this case, the tricky part was that our current spec setup does not support setting up multiple gem servers when some of them support new APIs and some of them don't. But I will give it another try.

Regarding a testing package registry,  I guess it could make sense although it feels slightly overkill for now given how little use we make of this realworld test harness.

Comment (Fryguy) [reply bundler/spec/realworld/double_check_spec.rb:45]: That's fair. As the owner of that gem server, we have no intention of removing anything, since we use that for production builds, so for now it's fine, but who knows what could happen in the future.

Thanks for the quick fix!

Comment (deivid-rodriguez) [reply bundler/spec/realworld/double_check_spec.rb]: I gave it another try and I managed to write a better spec, it should be ok now!
